modules/src/ics02_client/client_def.rs

```diff
@@ -216,8 +167,152 @@ pub enum AnyClient {
     Mock(MockClient),
 }
 
+impl AnyClient {
+    pub fn from_client_type(client_type: ClientType) -> AnyClient {
+        match client_type {
+            ClientType::Tendermint => Self::Tendermint(TendermintClient),
+
+            #[cfg(test)]
+            ClientType::Mock => Self::Mock(MockClient),
+        }
+    }
+}
+
+// ⚠️  Beware of the awful boilerplate below ⚠️
 impl ClientDef for AnyClient {
     type Header = AnyHeader;
     type ClientState = AnyClientState;
     type ConsensusState = AnyConsensusState;
+
+    fn check_header_and_update_state(
+        &self,
+        client_state: AnyClientState,
+        header: AnyHeader,
+    ) -> Result<(AnyClientState, AnyConsensusState), Box<dyn std::error::Error>> {
+        match self {
+            Self::Tendermint(client) => {
+                let (client_state, header) = downcast!(
+                    client_state => AnyClientState::Tendermint,
+                    header => AnyHeader::Tendermint,
+                )
+                .ok_or_else(|| error::Kind::ClientArgsTypeMismatch(ClientType::Tendermint))?;
+
+                let (new_state, new_consensus) =
+                    client.check_header_and_update_state(client_state, header)?;
+
+                Ok((
+                    AnyClientState::Tendermint(new_state),
+                    AnyConsensusState::Tendermint(new_consensus),
+                ))
+            }
+
+            #[cfg(test)]
+            Self::Mock(client) => {
+                let (client_state, header) = downcast!(
+                    client_state => AnyClientState::Mock,
+                    header => AnyHeader::Mock,
+                )
+                .ok_or_else(|| error::Kind::ClientArgsTypeMismatch(ClientType::Mock))?;
+
+                let (new_state, new_consensus) =
+                    client.check_header_and_update_state(client_state, header)?;
+
+                Ok((
+                    AnyClientState::Mock(new_state),
+                    AnyConsensusState::Mock(new_consensus),
+                ))
+            }
+        }
+    }
+
+    fn verify_client_consensus_state(
+        &self,
+        client_state: &Self::ClientState,
+        height: Height,
+        prefix: &CommitmentPrefix,
+        proof: &CommitmentProof,
+        client_id: &ClientId,
+        consensus_height: Height,
+        expected_consensus_state: &AnyConsensusState,
+    ) -> Result<(), Box<dyn std::error::Error>> {
+        match self {
+            Self::Tendermint(client) => {
+                let (client_state, expected_consensus_state) = downcast!(
+                    client_state => AnyClientState::Tendermint,
+                    expected_consensus_state => AnyConsensusState::Tendermint,
```

Comment:
The `expected_consensus_state` should be of the same type as the local chain (`Tendermint` in production or `Mock` in testing). This is different than the `ClientDef` type. For example`self` could be a client for `Substrate` while the `expected_consensus_state` stored by "our" client on the substrate chain is `Tendermint`. Not entirely sure of the best way to get the local chain info. One possibility is to pass the `Reader` context as a parameter and get the local chain type from there. Didn't have a chance to review this before it was merged, only looked at it now while merging master to `anca/proto_update`. Here we have a similar problem for the new `verify_full_client_state()` method. Let's try to take care of this in https://github.com/informalsystems/ibc-rs/pull/227